packages/langium/src/grammar/typification/type-validator.ts

```diff
@@ -0,0 +1,142 @@
+/******************************************************************************
+ * Copyright 2022 TypeFox GmbH
+ * This program and the accompanying materials are made available under the
+ * terms of the MIT License, which is available in the project root.
+ ******************************************************************************/
+
+import { AbstractRule, Grammar } from '../generated/ast';
+import { getRuleType } from '../grammar-util';
+import { MultiMap } from '../../utils/collections';
+import { collectDeclaredTypes } from './declared-types';
+import { collectInferredTypes } from './inferred-types';
+import { AstTypes, collectAllAstResources, Field, FieldType, InterfaceType, typeFieldToString, TypeType } from './types-util';
+
+export type TypeInconsistency = {
+    nodes: readonly AbstractRule[];
+    inconsistencyReasons: string[];
+}
+
+export function validateTypes(grammar: Grammar): TypeInconsistency[] {
+    const result: TypeInconsistency[] = [];
+    const validationResources = collectValidationResources(grammar);
+    for (const [inferredTypeName, inferredType] of validationResources.nameToInferredType.entries()) {
+        const declaredType = validationResources.nameToDeclaredType.get(inferredTypeName);
+        if (!declaredType) continue;
+
+        const inconsistencyReasons: string[] = [];
+        if (isType(declaredType) && isType(inferredType.type)) {
+            inconsistencyReasons.push(...checkAlternativesConsistency(inferredType.type.alternatives, declaredType.alternatives));
+        } else if (isInterface(declaredType) && isInterface(inferredType.type)) {
+            inconsistencyReasons.push(...checkFieldConsistency(inferredType.type.fields, declaredType.fields));
+            inconsistencyReasons.push(...checkStringElementConsistency(inferredType.type.superTypes, declaredType.superTypes, 'super type'));
+        } else {
+            inconsistencyReasons.push(`Inferred and declared versions of type ${inferredTypeName} have to be both types or interfaces.`);
+        }
+
+        if (inconsistencyReasons.length > 0) {
+            result.push({
+                nodes: inferredType.rules,
```

Comment:
Errors related to assignments should be shown on the respective assignment, not the containing grammar rule.